src/sentry/services/hybrid_cloud/rpc.py

```diff
@@ -0,0 +1,164 @@
+        service class that declares (but does not implement) at least one such method.
+
+        This method would return false, for example, for the local, database-backed
+        implementation that inherits from the base service. It also would return
+        false on an intermediate, abstract subclass that is meant to be extended to
+        declare other base services.
+        """
+
+        try:
+            next(cls._get_all_abstract_rpc_methods())
+        except StopIteration:
+            return False
+        else:
+            return True
+
+    @classmethod
+    @abstractmethod
+    def get_local_implementation(cls) -> "RpcService":
+        """Return a service object that runs locally.
+
+        The returned service object is (generally) the database-backed instance that
+        is called when we either receive a remote call from outside, or want to call
+        it within the same silo.
+
+        A base service class generally should override this class method, making a
+        forward reference to its own database-backed subclass.
+        """
+
+        raise NotImplementedError
+
+    @classmethod
+    def _create_local_delegation(cls) -> "RpcService":
+        """Create a lazy wrapper around the local implementation.
+
+        The indirection is necessary so that a base service's
+        `get_local_implementation` body can safely make a forward reference that
+        `resolve_to_delegation` won't call yet.
+        """
+        impl_attrname = "__cached_local_implementation"
+
+        @cached_property  # type: ignore
+        def cached_local_implementation(service_obj: "RpcService") -> "RpcService":
+            return service_obj.get_local_implementation()
+
+        def create_delegating_method(base_method: Callable[..., Any]) -> Callable[..., Any]:
+            method_name = base_method.__name__
+
+            def delegating_method(service_obj: "RpcService", **kwargs: Any) -> Any:
+                impl = getattr(service_obj, impl_attrname)
+                method = getattr(impl, method_name)
+                return method(**kwargs)
+
+            return delegating_method
+
+        overrides = {
+            service_method.__name__: create_delegating_method(service_method)
+            for service_method in cls._get_all_abstract_rpc_methods()
+        }
+        overrides[impl_attrname] = cached_local_implementation
+        remote_service_class = type(f"{cls.__name__}__LocalDelegate", (cls,), overrides)
+        return cast(RpcService, remote_service_class())
+
+    @classmethod
+    def _create_remote_delegation(cls) -> "RpcService":
+        """Create a service object that makes remote calls to another silo."""
+
+        def create_remote_method(base_method: Callable[..., Any]) -> Callable[..., Any]:
+            method_name = base_method.__name__
+
+            def remote_method(**kwargs: Any) -> Any:
+                # TODO: Handle kwarg serialization
```

Comment:
This depends on the code in #44242. I'll be merging the two branches to build this out.